Added ability to reference global resources

rpapallas opened the pull request:
When a user needs to embed images across different pages, having duplicate images locally in each page's `images` directory is redundant. I wanted to add a gallery of the same images across different pages/sections of the website. I couldn't find a neat way to do this currently. It appears that the shortcode would only look locally in each page's resources.

I added an extra option, `globalMatch`, that will look for images under the `/assets` directory instead, leveraging the [global resources utility from Hugo](https://gohugo.io/functions/resources/match/). This to work, a user will need to place those “repeating” images under `/assets` (images could exist under a subdirectory under `assets`).

## Example

```
.
├── archetypes
├── assets
│   ├── css
│   └── images
│       └── something
│           ├── test1.png
│           ├── test1.png.meta
├── config.toml
├── content
│   ├── gallery
│   │   └── index.md
│   ├── news
│   │   ├── something
│   │   │   └── index.md
├── static
└── themes
```

content/gallery/index.md:

```
{{<gallery
        globalMatch="images/something/*"
        sortOrder="asec"
        rowHeight="150"
        margins="5"
        thumbnailResizeOptions="600x600 q90 Lanczos"
        showExif=true
        previewType="blur"
        embedPreview=true
        loadJQuery=true
>}}
```

content/news/something/index.md:

```
{{<gallery
        globalMatch="images/something/*"
        sortOrder="asec"
        rowHeight="150"
        margins="5"
        thumbnailResizeOptions="600x600 q90 Lanczos"
        showExif=true
        previewType="blur"
        embedPreview=true
        loadJQuery=true
>}}
```

Both pages would now embed the same images, yet those images exist only once on the server.

I hope this is useful to others.

Rafael

mfg92 commented:
This looks like a promising feature to me. I will have a closer look at it and test it in the next few days.

In the meantime, it seems to me that the indentation here is a bit broken:
![grafik](https://github.com/user-attachments/assets/9cde2964-6ffd-4e4a-9adf-fb843adb0222)

rpapallas commented:
That's possibly because the file uses tabs for indentation and my editor uses spaces. Would you like to keep the tabs? If you don't mind using spaces, I can convert the entire file to spaces (usually more robust and avoids such issues).

mfg92 commented:
Thanks for the offer, but I prefer to stay with tabs. Not that I care that much, but changing every line would disrupt the git history/blame feature.

mfg92 commented:
I have now tested it and it works fine. But there is one problem: If both parameters, match and globalMatch, are present, the images that belong to the match parameter are displayed, hover, the metadata is missing because of line 106-108.

Could you rebase your commit, use tabs as intention and fix this? Then I would be happy to merge this PR.

rpapallas commented:
@mfg92 I think I have fixed the bug. I **think** I have also fixed the indentation issue. Finally, I added some details about this new option in the README.

mfg92 commented on README.md:
````diff
@@ -113,19 +113,72 @@ The text of the buttons and the regex used to filter has to be specified in a JS
+>}}
+```
+
+Note the `globalMatch` parameter, and the path `images/something/*` relative
````
This sentence needs a verb

rpapallas replied:
The verb 'Note' applies to both parts of the sentence: 'Note the global... **and** (note) the relative path...' 

I believe the sentence is clear as it stands. However, feel free to change it yourself if you prefer a more explicit phrasing.

mfg92 commented on layouts/shortcodes/gallery.html:
```diff
@@ -94,10 +99,15 @@
+
+			{{/* Get metadata from sidecar file, if present. Else an empty dictionary is used. */}}
 			{{ $metaFileName := print $original.Name ".meta"}}
 			{{ $metadata := $currentPage.Page.Resources.GetMatch ($metaFileName) }}
```
If $globalMatch is present, this line does unnecessary work. I think the solution is an if ... else ...

rpapallas replied:
Thanks for raising this. I used to have it with if and else, but for some reason I couldn't get it to work. Let me have another look.

rpapallas replied:
@mfg92 This is now fixed.

mfg92 reviewed (changes requested):
Thanks for the fix and the README update. I do have some feedback for you though.

mfg92 commented:
This is now part of of [v1.3.0](https://github.com/mfg92/hugo-shortcode-gallery/releases/tag/v1.3.0). Thanks a lot for your contribution.

rpapallas commented:
No problem; thank you very much for maintaining a useful shortcode.
